### ISSUE TITLE
Specify filename (not just directory) when downloading GDrive file

### ISSUE THREAD

**Issue by cspotcode** [scrthq/PSGSuite#188]
**Is your feature request related to a problem? Please describe.**

I have Google Drive files that contain forward slashes in their filename (not their path, actually in the filename itself)  I want to download the contents of such a file, but PSGSuite is trying to create a local file with slashes in the filename.  This fails because it expects to traverse into a subdirectory that does not exist.

**Describe the solution you'd like**

I want to be able to specify the full, local filename to download to.  This allows me to bypass any issues with the GDrive filename.  For example:

```
Get-GSDriveFile -FileID $idOfFileWithWeirdFilename -OutFilenamePath /foo/bar/normal-filename.zip
```

It will download the contents of the remote file and save it as `baz.zip`, regardless of the filename on GDrive.

**Describe alternatives you've considered**

**Additional context**

Here's the error I get.  The filename (name, not path) on Google Drive is `backup__2019-05-19T04:24:43Z__Personal-dev/@creationix/nvm.zip`  The name containers forward slash characters, which GDrive allows.

```
$ Get-GSDriveFile -FileId $id -OutFilePath '/tmp/157ee048-e182-4e13-94b5-abf0b1d3cd87'
Get-GSDriveFile : Exception calling "Create" with "1" argument(s): "Could not find a part of the path '/tmp/157ee048-e182-4e13-94b5-abf0b1d3cd87/backup__2019-05-19T04:24:43Z__Personal-dev/@creationix/nvm.zip.zip'."
```

**Comment by WJurecki** [comment]
The invalid filename part of this issue could be addresses as you did in #169.

While I could also find specifying a desired filename would be beneficial, I don't know how to suggest to reconcile this filename option with the ability to send an array of File IDs to download.

**Comment by cspotcode** [comment]
#169 sounds like it will change the target filename to be safe, but will not tell me what that target filename is.  So I have to do some manual string manipulation and hope it matches the behavior of PSGSuite.  I'm writing an automated backup and restore script, so immediately after downloading a file, I need to pass it to `unzip` and then delete it.  The filename's arbitrary, but I need to have easy access to it.

For sending an array of FileIDs, can I simply specify an array of local filenames to match, with a one-to-one relationship?  Or simply disallow multiple FileIDs when I supply a target filename?  So if I want to download multiple FileIDs to specific local filenames, I'll need to do it in a `foreach()` loop.

**Comment by WJurecki** [comment]
@cspotcode, I see the issue you would have with not knowing the output filename.  I would as well in my intended process.

How about a property added to the output object to provide the actual file name used to output the file?

In this case the output object would have a `name` property with the original/Google Drive name, in this case, a value of `backup__2019-05-19T04:24:43Z__Personal-dev/@creationix/nvm.zip` and the new `OutputName` property with a value of `backup__2019-05-19T04_24_43Z__Personal-dev_@creationix_nvm.zip`

@scrthq I hope this discussion helps.  I'm not trying to dictate any design ideas, but rather come up with a potential solution that made fit more needs.  Your choice!

**Comment by cspotcode** [comment]
That would work for me.

Today I generate a temp filename using OS APIs, download to that filename, then delete the temp file when I'm done.  (I've patched my local copy of PSGSuite)

With your proposed solution, I would instead generate a new temp directory name, create the new empty directory, and download the file into that new directory.  Then I would delete my temp directory.  This would be necessary to avoid any possible name conflicts with other files in the OS's temp directory.

On Linux, `mktemp` can generate either files or directories, so both approaches are easy.  On Windows, I'm not sure there's a straightforward OS-provided way to safely create temp directories, other than creating it in the system's temp directory with a randomly-generated name.

**Comment by scrthq** [comment]
@cspotcode - I appreciate the feedback! How does expanding the `OutFilePath` parameter to allow full paths to be specified, with it falling back to using a *special-character-replaced* version of the file name as is in Drive?

To satisfy the resulting FileName piece, I can also add in the final `OutFilePath` value to the returned object as well.

***

Sample code to test the special-character-replacement against a string of your own (feedback appreciated in case there are characters missed!):

```powershell
$Name = 'backup__2019-05-19T04:24:43Z__Personal-dev/@creationix/nvm.zip'
$Name -replace "[$(([System.IO.Path]::GetInvalidFileNameChars() + [System.IO.Path]::GetInvalidPathChars()) -join '')]","_"
```

Resulting output on Windows 10: `backup__2019-05-19T04_24_43Z__Personal-dev_@creationix_nvm.zip`

***

Sample returned object when downloading something to a specified path:

![image](https://user-images.githubusercontent.com/12724445/58306266-d9e62c00-7daf-11e9-8439-25904586dabe.png)

***

@WJurecki - Thanks for jumping in here!!

**Comment by scrthq** [comment]
@cspotcode - With the above adjustment, existing usage could continue to function as is and you also have the option of piping in filenames and capturing the `OutFilePath` value of the resulting object as well as specifying it per file via loop.

**Comment by scrthq** [comment]
Hey @cspotcode - This has been deployed in v2.28.0! Let me know if all is well or if you are seeing any issues!

**Comment by WJurecki** [comment]
@scrthq - This appears to mostly work.

There is a problem in your private `Get-SafeFileName`

Consider this:

![image](https://user-images.githubusercontent.com/48518449/58735991-e5de7900-83c1-11e9-808a-3cca701384b8.png)

the \ still made it into the 'clean name' because the \ in the replace command needs to be escaped and actually \\\\ to work properly.

![image](https://user-images.githubusercontent.com/48518449/58736161-77e68180-83c2-11e9-9d7c-51806f7e3f53.png)


Also, you should probably consider not only the name needs to be cleaned, but also the extension.

consider these perfectly valid Drive filenames
![image](https://user-images.githubusercontent.com/48518449/58736121-4e2d5a80-83c2-11e9-957e-b2634f357a52.png)

Sorry for the bad news.  :-(

**Comment by scrthq** [comment]
Hey @WJurecki - Fair! The escaping is done against the current OS's invalid file and path characters, leaving anything alphanumeric or `[System.IO.Path]::DirectorySeparatorChar` alone. On Windows, `[System.IO.Path]::DirectorySeparatorChar -eq '\'`, so that's why that's left alone after being returned from `Get-SafeFileName`.

I definitely see your point though in that the files in Drive are just that, files, so it would make sense to also replace any valid path separators so the resulting file actually is a file and not something that got buried in an unexpected naming path. Testing it out actually returns an error since it's trying to access a path that doesn't exist due to the DirectorySeparatorChar being in the FileName:

![image](https://user-images.githubusercontent.com/12724445/58736951-4a4f0780-83c5-11e9-9511-ade679b95b00.png)

**Comment by WJurecki** [comment]
@scrthq 

but `[System.IO.Path]::GetInvalidFileNameChars()` returns `"<>|:*?\/` (along with some non-printable characters) and `[System.IO.Path]::GetInvalidPathChars()` returns `"<>|` 

I really appears that, at least on Windows, `-replace` needs `\` to be escaped as `\\` to work correctly.

in the code `"[$(([System.IO.Path]::GetInvalidFileNameChars() + [System.IO.Path]::GetInvalidPathChars()) -join '')]"` results in `["<>|:*?\/"<>|]` (non-printable characters removed for display) when it really needs to be `["<>|:*?\\/"<>|]` to work correctly.

**Comment by scrthq** [comment]
Aha! This should do the trick instead, just wrapped the resulting character string in `[RegEx]::Escape()` so that special RegEx chars like `\` are treated as literal. I wanted to avoid explicitly calling out chars to replace in the event that specific chars are acceptable depending on the OS.

Test block:

```powershell
'Test\File/|Name' -replace "[$([RegEx]::Escape("$(([System.IO.Path]::GetInvalidFileNameChars() + [System.IO.Path]::GetInvalidPathChars()) -join '')"))]","_"
```

Thoughts @WJurecki ?

**Comment by WJurecki** [comment]
@scrthq, that looks great!

**Comment by scrthq** [comment]
@WJurecki fixed rolled out in 2.28.1!
